Not open on Android 16

### What is the bug or the crash? What were your expectations and what actually happened?

Force close on open 
https://bin.mkr.pw/~6a2625111f60ed58f17e5d15

time: 1780884748643
msg: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.io.File.getAbsolutePath()' on a null object reference
stacktrace: java.lang.RuntimeException: Unable to start activity ComponentInfo{ch.opengis.qfield/ch.opengis.qfield.QFieldActivity}
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4496)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:4699)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:224)
	at android.app.servertransaction.TransactionExecutor.executeNonLifecycleItem(TransactionExecutor.java:133)
	at android.app.servertransaction.TransactionExecutor.executeTransactionItems(TransactionExecutor.java:103)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:80)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2961)
	at android.os.Handler.dispatchMessage(Handler.java:132)
	at android.os.Looper.dispatchMessage(Looper.java:333)
	at android.os.Looper.loopOnce(Looper.java:263)
	at android.os.Looper.loop(Looper.java:367)
	at android.app.ActivityThread.main(ActivityThread.java:9282)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:566)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:929)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.io.File.getAbsolutePath()' on a null object reference
	at ch.opengis.qfield.QFieldActivity.prepareQtActivity(QFieldActivity.java:508)
	at ch.opengis.qfield.QFieldActivity.onCreate(QFieldActivity.java:153)
	at android.app.Activity.performCreate(Activity.java:9306)
	at android.app.Activity.performCreate(Activity.java:9284)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1547)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:4480)
	... 14 more


### Steps to reproduce the issue

Just force close on open

### Version

4.2.1

### Operating system name

Android

### Operating system version

Android 16

### Reinstall QField

- [x] I have a fresh install of the latest QField version, but the problem persists.
- [ ] Problem can be reliably reproduced, doesn't happen randomly.
- [ ] Problem happens with all files and projects, not only some files or projects.

### Additional context

_No response_

--- a/platform/android/src/ch/opengis/qfield/QFieldActivity.java
+++ b/platform/android/src/ch/opengis/qfield/QFieldActivity.java
@@ -504,7 +504,8 @@ public class QFieldActivity extends QtActivity {
         for (File file : externalFilesDirs) {
             if (file != null) {
                 // Don't duplicate primary external files directory
-                if (file.getAbsolutePath().equals(
+                if (primaryExternalFilesDir != null &&
+                    file.getAbsolutePath().equals(
                         primaryExternalFilesDir.getAbsolutePath())) {
                     continue;
                 }

--- a/platform/android/src/ch/opengis/qfield/QFieldActivity.java
+++ b/platform/android/src/ch/opengis/qfield/QFieldActivity.java
@@ -268,7 +268,7 @@ public class QFieldActivity extends QtActivity {
                 String filePath = QFieldUtils.getPathFromUri(context, uri);
                 String importDatasetPath = "";
                 String importProjectPath = "";
-                File externalFilesDir = getExternalFilesDir(null);
+                File externalFilesDir = getApplicationDir();
                 if (externalFilesDir != null) {
                     importDatasetPath = externalFilesDir.getAbsolutePath() +
                                         "/Imported Datasets/";
@@ -463,7 +463,8 @@ public class QFieldActivity extends QtActivity {
 
         List<String> dataDirs = new ArrayList<String>();
 
-        File primaryExternalFilesDir = getExternalFilesDir(null);
+        File primaryExternalFilesDir = getApplicationDir();
+
         if (primaryExternalFilesDir != null) {
             String dataDir = primaryExternalFilesDir.getAbsolutePath() + "/";
             // create import and creation directories
@@ -569,10 +570,12 @@ public class QFieldActivity extends QtActivity {
     }
 
     private String getApplicationDirectory() {
-        File primaryExternalFilesDir = getExternalFilesDir(null);
+        File primaryExternalFilesDir = getApplicationDir();
+
         if (primaryExternalFilesDir != null) {
             return primaryExternalFilesDir.getAbsolutePath();
         }
+
         return "";
     }
 
@@ -590,14 +593,14 @@ public class QFieldActivity extends QtActivity {
                 Environment.getExternalStorageDirectory();
         }
 
-        File primaryExternalFilesDir = getExternalFilesDir(null);
-
+        File primaryExternalFilesDir = getApplicationDir();
         File[] externalFilesDirs = getExternalFilesDirs(null);
         for (File file : externalFilesDirs) {
             if (file != null) {
                 // Don't duplicate external files directory or storage
                 // path already added
-                if (file.getAbsolutePath().equals(
+                if (primaryExternalFilesDir != null &&
+                    file.getAbsolutePath().equals(
                         primaryExternalFilesDir.getAbsolutePath())) {
                     continue;
                 }
@@ -1005,7 +1008,8 @@ public class QFieldActivity extends QtActivity {
     }
 
     void importDatasets(Uri[] datasetUris) {
-        File externalFilesDir = getExternalFilesDir(null);
+        File externalFilesDir = getApplicationDir();
+
         if (externalFilesDir == null || datasetUris.length == 0) {
             return;
         }
@@ -1061,7 +1065,8 @@ public class QFieldActivity extends QtActivity {
     }
 
     void importProjectFolder(Uri folderUri) {
-        File externalFilesDir = getExternalFilesDir(null);
+        File externalFilesDir = getApplicationDir();
+
         if (externalFilesDir == null) {
             return;
         }
@@ -1105,7 +1110,8 @@ public class QFieldActivity extends QtActivity {
     }
 
     void importProjectArchive(Uri archiveUri) {
-        File externalFilesDir = getExternalFilesDir(null);
+        File externalFilesDir = getApplicationDir();
+
         if (externalFilesDir == null) {
             return;
         }
@@ -1172,7 +1178,7 @@ public class QFieldActivity extends QtActivity {
     }
 
     void updateProjectFromArchive(Uri archiveUri) {
-        File externalFilesDir = getExternalFilesDir(null);
+        File externalFilesDir = getApplicationDir();
         if (externalFilesDir == null) {
             return;
         }
@@ -1332,6 +1338,16 @@ public class QFieldActivity extends QtActivity {
         }
     }
 
+    private File getApplicationDir() {
+        File applicationDirectory = getExternalFilesDir(null);
+        if (applicationDirectory == null) {
+            // On some Android devices, getExternalFilesDir(null) can return a
+            // null value, fallback to getFilesDir()
+            applicationDirectory = getFilesDir();
+        }
+        return applicationDirectory;
+    }
+
     protected void onActivityResult(int requestCode, int resultCode,
                                     Intent data) {
         if (requestCode == CAMERA_RESOURCE) {
@@ -1444,7 +1460,7 @@ public class QFieldActivity extends QtActivity {
         } else if (requestCode == IMPORT_DATASET &&
                    resultCode == Activity.RESULT_OK) {
             Log.d("QField", "handling import dataset(s)");
-            File externalFilesDir = getExternalFilesDir(null);
+            File externalFilesDir = getApplicationDir();
             if (externalFilesDir == null || data == null) {
                 return;
             }
@@ -1509,7 +1525,7 @@ public class QFieldActivity extends QtActivity {
         } else if (requestCode == IMPORT_PROJECT_FOLDER &&
                    resultCode == Activity.RESULT_OK) {
             Log.d("QField", "handling import project folder");
-            File externalFilesDir = getExternalFilesDir(null);
+            File externalFilesDir = getApplicationDir();
             if (externalFilesDir == null || data == null) {
                 return;
             }
@@ -1548,7 +1564,7 @@ public class QFieldActivity extends QtActivity {
         } else if (requestCode == IMPORT_PROJECT_ARCHIVE &&
                    resultCode == Activity.RESULT_OK) {
             Log.d("QField", "handling import project archive");
-            File externalFilesDir = getExternalFilesDir(null);
+            File externalFilesDir = getApplicationDir();
             if (externalFilesDir == null || data == null) {
                 return;
             }
@@ -1596,7 +1612,7 @@ public class QFieldActivity extends QtActivity {
         } else if (requestCode == UPDATE_PROJECT_FROM_ARCHIVE &&
                    resultCode == Activity.RESULT_OK) {
             Log.d("QField", "handling updating project from archive");
-            File externalFilesDir = getExternalFilesDir(null);
+            File externalFilesDir = getApplicationDir();
             if (externalFilesDir == null || data == null) {
                 return;
             }

